src/state/hooks/hook-caller.js

```diff
@@ -0,0 +1,311 @@
+// @flow
+import messagePreset from './message-preset';
+import type { HookCaller } from './hooks-types';
+import type {
+  Announce,
+  Hooks,
+  HookProvided,
+  State as AppState,
+  DragState,
+  DragStart,
+  DragUpdate,
+  DropResult,
+  DraggableLocation,
+  DraggableDescriptor,
+  DroppableDimension,
+  OnDragStartHook,
+  OnDragUpdateHook,
+  OnDragEndHook,
+} from '../../types';
+
+type State = {
+  isDragging: boolean,
+  start: ?DraggableLocation,
+  lastDestination: ?DraggableLocation,
+  hasMovedFromStartLocation: boolean,
+}
+
+type AnyHookFn = OnDragStartHook | OnDragUpdateHook | OnDragEndHook;
+type AnyHookData = DragStart | DragUpdate | DropResult;
+
+const notDragging: State = {
+  isDragging: false,
+  start: null,
+  lastDestination: null,
+  hasMovedFromStartLocation: false,
+};
+
+const areLocationsEqual = (current: ?DraggableLocation, next: ?DraggableLocation) => {
+  // if both are null - we are equal
+  if (current == null && next == null) {
+    return true;
+  }
+
+  // if one is null - then they are not equal
+  if (current == null || next == null) {
+    return false;
+  }
+
+  // compare their actual values
+  return current.droppableId === next.droppableId &&
+    current.index === next.index;
+};
+
+const getAnnouncerForConsumer = (announce: Announce) => {
+  let wasCalled: boolean = false;
+  let isExpired: boolean = false;
+
+  // not allowing async announcements
+  setTimeout(() => {
+    isExpired = true;
+  });
+
+  const result = (message: string): void => {
+    if (wasCalled) {
```

Comment:
we do some guarding here to prevent double execution of `announce` as well as preventing async `announce` calls